src/gcp_scanner/scanner.py

```diff
@@ -129,166 +130,104 @@ def save_results(res_data: Dict, res_path: str, is_light: bool):
-          # this path is used by the storage bucket crawler as of now.
-          crawler_config['gcs_output_path'] = gcs_output_path
-          # crawl the data
-          crawler = CrawlerFactory.create_crawler(crawler_name)
-          client = ClientFactory.get_client(client_name).get_service(
-            credentials,
-          )
-          project_result[crawler_name] = crawler.crawl(
-            project_id,
-            client,
-            crawler_config,
-          )
-
-      # Call other miscellaneous crawlers here
-      if is_set(scan_config, 'gke_clusters'):
-        gke_client = gke_client_for_credentials(credentials)
-        project_result['gke_clusters'] = misc_crawler.get_gke_clusters(
-          project_id,
-          gke_client,
-        )
-      if is_set(scan_config, 'gke_images'):
-        project_result['gke_images'] = misc_crawler.get_gke_images(
-          project_id,
-          credentials.token,
-        )
-
-      # Iterate over discovered service accounts by attempting impersonation
-      project_result['service_account_edges'] = []
-      updated_chain = chain_so_far + [sa_name]
 
-      if scan_config is not None:
-        impers = scan_config.get('service_accounts', None)
-      else:
-        impers = {'impersonate': False}  # do not impersonate by default
-
-      # trying to impersonate SAs within project
-      if impers is not None and impers.get('impersonate', False) is True:
-        iam_client = iam_client_for_credentials(credentials)
-        if is_set(scan_config, 'iam_policy') is False:
-          iam_policy = CrawlerFactory.create_crawler('iam_policy').crawl(
-            project_id,
-            ClientFactory.get_client('cloudresourcemanager').get_service(
-              credentials,
-            ),
-          )
-
-        project_service_accounts = get_sas_for_impersonation(iam_policy)
-        for candidate_service_account in project_service_accounts:
-          try:
-            logging.info('Trying %s', candidate_service_account)
-            creds_impersonated = credsdb.impersonate_sa(
-              iam_client, candidate_service_account)
-            context.service_account_queue.put(
-              (candidate_service_account, creds_impersonated, updated_chain))
-            project_result['service_account_edges'].append(
-              candidate_service_account)
-            logging.info('Successfully impersonated %s using %s',
-                         candidate_service_account, sa_name)
-          except Exception:
-            logging.error('Failed to get token for %s',
-                          candidate_service_account)
-            logging.error(sys.exc_info()[1])
-
-      logging.info('Saving results for %s into the file', project_id)
-
-      save_results(sa_results, output_path, light_scan)
-      # Clean memory to avoid leak for large amount projects.
-      sa_results.clear()
+  # Iterate over discovered service accounts by attempting impersonation
```

Comment:
I'd prefer that functionality to be implemented in this commit instead of just commenting it out. What we can basically do is to create a separate function that does impersonation and then call it from the loop starting at line 393. This way your build a list of available SAs and associated projects and scan them.

Reply:
So, we basically run two loops, first for the SA impersonation and the other for crawling.

Reply:
Done